### DISCUSSION TITLE
Element implicitly has an 'any' type because type 'typeof globalThis' has no index signature.

### DISCUSSION THREAD

**Discussion by SamSverko** [prisma/prisma#10037] [Q&A] [answered]
I have this `./lib/prisma.ts` file copied from this [Vercel + Prisma tutorial](https://vercel.com/guides/nextjs-prisma-postgres#step-3.-install-and-generate-prisma-client):

```ts
import { PrismaClient } from '@prisma/client';

let prisma: PrismaClient;

if (process.env.NODE_ENV === 'production') {
  prisma = new PrismaClient();
} else {
  if (!global.prisma) {
    global.prisma = new PrismaClient();
  }
  prisma = global.prisma;
}

export default prisma;
```

I am using TypeScript, and get this error for the `global.prisma` instances:

```
Element implicitly has an 'any' type because type 'typeof globalThis' has no index signature.ts(7017)
```

Any help understanding or removing this error would be deeply appreciated. Thank you.

**Comment by aqrln** [comment, marked as answer]
Hi @SamSverko! I think changing it like this might help:

```ts
import { PrismaClient } from '@prisma/client';

let prisma: PrismaClient;

if (process.env.NODE_ENV === "production") {
  prisma = new PrismaClient();
} else {
  let globalWithPrisma = global as typeof globalThis & {
    prisma: PrismaClient;
  };
  if (!globalWithPrisma.prisma) {
    globalWithPrisma.prisma = new PrismaClient();
  }
  prisma = globalWithPrisma.prisma;
}

export default prisma;
```

**Reply by SamSverko** [reply]
Oh thank you! That makes the `global.prisma` error go away!

However, TS does not like the use of `var` (`Unexpected var, use let or const instead.`). When I change it to either be `let` or `const`, the `global.prisma` error becomes:

```
Property 'prisma' does not exist on type 'typeof globalThis'.
```

**Reply by aqrln** [reply]
> TS does not like the use of var (Unexpected var, use let or const instead.)

Is it really TypeScript or your linter? I just tried it and it seems to work for me. Besides, this error message looks very similar to what `no-var` rule in ESLint produces.

**Reply by aqrln** [reply]
Either way, this solution might not be such a good idea after all since it will leak this global declaration to other files in your project, which will make it possible to use it without importing, breaking the app in production.

It's better to do something like this:

```ts
import { PrismaClient } from '@prisma/client';

let prisma: PrismaClient;

if (process.env.NODE_ENV === "production") {
  prisma = new PrismaClient();
} else {
  let globalWithPrisma = global as typeof globalThis & {
    prisma: PrismaClient;
  };
  if (!globalWithPrisma.prisma) {
    globalWithPrisma.prisma = new PrismaClient();
  }
  prisma = globalWithPrisma.prisma;
}

export default prisma;
```

**Reply by aqrln** [reply]
I've updated the answer as well

**Reply by SamSverko** [reply]
Ah, yes. That was an ESLint error, my mistake! Thank you for the explanation and fix!

**Reply by bagelbomb** [reply]
I think this would be a little bit better implementation because it accomplishes the same thing but is also more straightforward and avoids [exporting mutable variables](https://next.sonarqube.com/sonarqube/coding_rules?open=typescript%3AS6861&rule_key=typescript%3AS6861) (apologies for going slightly off-topic):
```typescript
import { PrismaClient } from '@prisma/client';

const globalWithPrisma = global as typeof globalThis & { prisma: PrismaClient };

if (process.env.NODE_ENV !== 'production' && !globalWithPrisma.prisma) {
  globalWithPrisma.prisma = new PrismaClient();
}

const prisma = process.env.NODE_ENV === 'production' ? new PrismaClient() : globalWithPrisma.prisma;

export default prisma;

```

**Reply by DylanPrimera** [reply]
This works well for me. Thanks!

**Comment by shawnmclean** [comment]
Can someone explain this code? Why do we need to put this on global?

**Reply by alex067** [reply]
It is to resolve this issue: https://www.prisma.io/docs/guides/database/troubleshooting-orm/help-articles/nextjs-prisma-client-dev-practices

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
